### ISSUE TITLE
discussion: separate out badge info from attendee info

### ISSUE THREAD

**Issue by binary1230** [magfest/ubersystem#1351]
This came up in an unrelated discussion about barcodes and dealing with lost badges.  This is a big discussion.

The idea is that we could/should separate out attendee information from badge information, for two reasons:

1) the abstraction is getting kind of leaky, and the Attendee table has grown de-normalized.  while the mapping for people => badges is 1:1 for most cases, it isn't always.

examples: 
* if you lose a badge, we currently (I believe, correct me if I'm wrong) mark the Attendee as "LOST", and create a new Attendee that is the replacement badge.  There's now the same person twice in the system.  We didn't lose the attendee, we lost the badge.
* someone wants a staff and supporter badge, or a guest badge and a supporter badge

2) This lends itself more naturally to a persistent ubersystem, where Attendees persist through all years, and their badge entries gain them access to each event. 
i.e. one Attendee would hold multiple Badges for each of the years.  they're a guest for m14, staff for m13, and supporter for magclassic.

We could split out the Badge-specific stuff into it's own model/table like this:

Attendee: just stuff about *the person* and nothing about their *badges*
- attendee_id (primary key, unique and independent of magfest event like magstock3, m14, etc)
- first name, last name, emergency contact info, etc
- person UUID (never changes across all magfest events)
- admin notes

Badge:
- owner attendee_id
- badge_id (NOT badge number, primary key unique for all badges across all magfest events)
- badge_type (guest, attendee, supporter)
- badge_num 
- barcode_num
- badge_status (new, invalid, complete, etc, is lost)
- event_id (m14, magclassic, magstock)
- paid (yes, no, how much, discounts, etc)
- group mapping info, is this part of a group, etc
- ribbon

In a "lost badge" scenario, we'd mark their current Badge as "LOST", and assign a new badge to them by adding a new Badge instance to the system and mapping it to their attendee.  That Attendee would have 2 Badges, one is lost and can't be used for stuff, and one is not lost.   if they bring their lost badge back because they found it, we remove the entry for the new badge and it goes back in the unusued pile of badges at reg.

We'd have to think carefully about Groups, though I think their definition becomes clear when you think about it in the context of multiple Events.  A group is a Group of Badges in a particular Event.

Mag8.5 group "fighting mongooses" = badges 7,12,76,94 (all of them 8.5 badges)
MagClassic group "Armcannon" = badges 9,5,1,8 (all of them magclassic badges)

**Comment by binary1230** [comment]
obviously this is a GIGANTIC refactor, so not proposing it anytime soon.  just, might be a good step towards getting us to persistent ubersystem, and also clear up some workflow issues with multiple badges, lost badges, etc.

**Comment by earl7399** [comment]
We do *NOT* mark the attendee as lost.  We record the badge # as lost, and
assign a new badge to the same attendee.  Victoria can give better details,
I suspect.  And of course correct me if I'm wrong.

--
Robert Earl <r.a.earl@gmail.com>
FAX: 1-440-919-5086

On Thu, Aug 6, 2015 at 11:06 AM, Dominic Cerquetti <notifications@github.com

**Comment by kitsuta** [comment]
Nah, last year we totally did create a new badge for the attendee. There's no way to record a badge _number_ as lost without an attendee attached to that number. Back when we simply changed badge numbers on a single attendee, you could no longer search for that badge number, creating much confusion.

Like I mentioned in the email thread, I'm not thrilled about the prospect of completely refactoring the core paradigm of uber/RAMS.

The main thing that worries me about it is the potential to open up workflows we find extremely undesirable; specifically, the PAX-style workflow of "one person owns 6 badges, gets to give them to whomever." Once a badge no longer needs a unique person attached to it, I can see people assuming that this is a correct workflow.

Even in cases we have now where people deserve two badges, we really don't _want_ people walking around with two valid badges. An extra badge is very easy to lend to a friend - we need to fix our workflow so that people don't have two, not make our code support it. Speaking of which, we totally already did that for Supporter badges. ;)

By the way, now that we have badge statuses, we can mark lost badges as invalid, which is the equivalent of deleting them. So that use-case is pretty much fixed already (though... we need to actually code that in.)

**Comment by earl7399** [comment]
Told you she could.

--
Robert Earl <r.a.earl@gmail.com>
FAX: 1-440-919-5086

**Comment by binary1230** [comment]
mroe discussion of this topic is here:
https://github.com/magfest/ubersystem/pull/1520

**Comment by kitsuta** [comment]
Revisiting time!

@EliAndrewC and I have done a lot of talking about this and I am more on-board with the idea than before.

Because this has a lot of tie-ins with multi-year functionality, our current plan is to do this in a separate Sideboard plugin. This plugin would likely _not_ be "RAMS" in that it would not import anything from the core RAMS plugin and could have a totally different structure. Heck, I'm even toying with the idea of making it a straight-up Django app.

We would use an API to have the Accounts plugin talk to the various events/years; the Accounts plugin would have all the 'person' data and each badge would have a 'cached' copy of that data which stops getting updated after the end of the event (this addresses a major user story of mine, 'I don't want a badge from an old event to have current personal information'). The cached copy would also make sure we always had a copy of the data even if we lost access to the Accounts server.

The main 'breaking change' here is that we would likely require the Accounts plugin for RAMS to work. Because it can be loaded on the same server as the core plugin, this should be relatively painless for events. And for MAGFest, we can run it on a different server entirely.

**Comment by binary1230** [comment]
random note, 

Discussion happening in reg channels about a use case we should think about

Who owns a badge, and what happens if a badge is transferred who is the new owner the person who paid for it or the person whose name it's in.  Happens a lot with couples breaking up and wanting to change ownership of badge pairs.  Also transfers

It's a complicated discussion with way more nuance than I'm doing justice to here, and one of the solutions gets simpler if we decouple attendees from badges.  We should check in with reg crew when doing this redesign
